### PR TITLE
Caching cmf values

### PR DESCRIPTION
**Pull request by rotelimelech**
Enabling to access the potential matrix through `cmf[x,y]`, and using cached data of the potential to save calculations

### CODE REVIEW

File: ramanujan/cmf/cmf.py [hunk 1]
```diff
@@ -13,7 +13,8 @@ class CMF:
     $Mx(x, y) \cdot My(x+1, y) = My(x, y) \cdot Mx(x, y+1)$
     """
 
-    def __init__(self, Mx: Matrix, My: Matrix):
+    def __init__(self, Mx: Matrix, My: Matrix, maximal_cache_dims=(50, 50), 
+        initial_mat=Matrix(sp.eye(2)), initial_loc=(1,1), potential_cache_file=None):
```
Comment (TotallyReal) [inline ramanujan/cmf/cmf.py:17]: There are a lot of new parameter, which are only required for the caching, and not the "standard" CMF construction. Consider putting them into a single object, so it won't be confusing

Comment (TotallyReal) [inline ramanujan/cmf/cmf.py:17]: There are a lot of new parameter, which are only required for the caching, and not the "standard" CMF construction. Consider putting them into a single object, so it won't be confusing

File: ramanujan/cmf/cmf.py [hunk 2]
```diff
@@ -28,9 +29,22 @@ def __init__(self, Mx: Matrix, My: Matrix):
         if simplify(Mxy - Myx) != Matrix([[0, 0], [0, 0]]):
             raise ValueError("The given Mx and My matrices are not conserving!")
 
+        self.initial_loc = initial_loc
```
Comment (TotallyReal) [inline ramanujan/cmf/cmf.py:32]: If you create this new "cache" object, you can use it here as well

File: ramanujan/cmf/cmf.py [hunk 3]
```diff
@@ -123,3 +137,68 @@ def limit(
             the walk multiplication as defined above.
         """
         return self.walk(trajectory, iterations, start).limit(vector)
+
+    def __getitem__(self, location):
```
Comment (TotallyReal) [inline ramanujan/cmf/cmf.py:141]: Add what is a "location" (I guess tuple\list of x,y?)

File: ramanujan/cmf/cmf.py [hunk 4]
```diff
@@ -123,3 +137,68 @@ def limit(
...
+        """
+        calculates the CMF in first quandrant up to (max_x, max_y).
+        """
+        
+        # The calculation method in `potential` takes steps over the x axis first, then over y.
+        # Since it caches elements while calculating, this will lead to caching of the entire space. 
+        for x_i in range(self.initial_loc[0], max_x+1):
+            self.potential([x_i, max_y])
+
+    def dump_potential_to_file(self, dest_path):
```
Comment (TotallyReal) [inline ramanujan/cmf/cmf.py:198]: These functions should probably be part of that new object as well. Namely, you should have something like cmf_object.cache_object.dump() \ load()

File: ramanujan/cmf/known_cmfs/known_cmfs.py [hunk 5]
```diff
@@ -1,5 +1,8 @@
 import sympy as sp
+import os
```
Comment (TotallyReal) [inline ramanujan/cmf/known_cmfs/known_cmfs.py:2]: keep it ordered - either all of the sympy imports before or all after the os imports

File: ramanujan/cmf/cmf_test.py [hunk 6]
```diff
@@ -1,8 +1,10 @@
+import sympy as sp
```
Comment (TotallyReal) [inline ramanujan/cmf/cmf_test.py:1]: put sympy imports together

File: ramanujan/cmf/cmf_test.py [hunk 7]
```diff
@@ -77,3 +79,14 @@ def test_substitute_trajectory_walk_equivalence():
     assert subbed.walk({n: 1}, iterations, {n: 1}) == unsubbed.walk(
         trajectory, iterations, start
     )
+
+
+def test_cache_calculation_method():
+    cmf = known_cmfs.zeta3()
+    cmf.reset_potential_cache([40,40], initial_mat=Matrix(sp.eye(2)))
+    assert cmf.walk({x:1, y:1}, 40).limit(Vector.zero()) == cmf[40,40].limit(Vector.zero())
```
Comment (TotallyReal) [inline ramanujan/cmf/cmf_test.py:87]: please add some comment about the difference between using "walk" and "[40,40]". It is not clear from the outside that one uses cache and the other not (or which one)

**Review by TotallyReal — CHANGES_REQUESTED** [review]

**Comment by TotallyReal** [comment]
Also, once you try to push the code, it is easier if you squash the commits together

File: ramanujan/cmf/cmf.py [hunk 8]
```diff
@@ -28,9 +29,22 @@ def __init__(self, Mx: Matrix, My: Matrix):
...
+        self.initial_loc = initial_loc
+        if potential_cache_file is not None:
+            self.potential_cache = self.load_potential_from_file(potential_cache_file)
+        else:
+            self.reset_potential_cache(maximal_cache_dims, initial_mat)
+
     def __repr__(self):
         return f"CMF({self.Mx}, {self.My})"
 
+    def reset_potential_cache(self, maximal_cache_dims, initial_mat):
```
Comment (TotallyReal) [inline ramanujan/cmf/cmf.py:41]: If you create this new "cache" object, you can use it here as well. This "reset" method, should probably be a method of this new object

File: ramanujan/cmf/cmf.py (hunk 1 above)
Comment (RotemKalisch) [reply ramanujan/cmf/cmf.py:17]: Agreed. Also, sometimes we'd like to load it after a parametric CMF has beed substituted, and it doesn't make sense yet to load it on the construction of the parametric CMF.

**Comment by RotemKalisch** [comment]
Closing this PR as it's highly outdated

**Pull request closed without merging**